### PR TITLE
Ver 4.0.0

### PR DESCRIPTION
**Pull request by shreyashah**
Integrate vxr_ssh changes, telnetlib changes, bsp, cloud_native, and other custom cafykit changes
bsp and cloud_native need to be tested.
Will elaborate this description more 
allure.cisco.com/ws/shreyash-sjc/cafyinfra/cafykit/work/archive/test_xr_methods_20230405-161958_p2580308/reports/index.html -- using vxr_sim 

Netmiko logger logs: /ws/shreyash-ott/netmiko.log

### CODE REVIEW

**Review by sangmeshcp — APPROVED** [review]

File: netmiko/cisco/cisco_xr.py [hunk 1]
```diff
@@ -39,6 +39,10 @@ def commit(
         label: str = "",
```
Comment (akshansh237) [inline netmiko/cisco/cisco_xr.py:64]: Generic comment: we discussed on creating new files and overriding methods in it, so that future netmiko upgrades are easier. did we drop that plan ?

File: netmiko/cisco/cisco_vxr_ssh.py [hunk 2]
```diff
@@ -0,0 +1,634 @@
...
+    formatter = logging.Formatter(fmt='%(asctime)s %(levelname)-8s %(message)s',
+                                  datefmt='%Y-%m-%d %H:%M:%S')
+    paramiko_log_folder = os.path.join(work_dir, 'test_paramiko.log')
+    paramiko.util.log_to_file(paramiko_log_folder, level="DEBUG")
+    fh = logging.FileHandler(paramiko_log_folder)
+    fh.setFormatter(formatter)
+    log = logging.getLogger("netmiko")
+    log.addHandler(fh)
+    log.setLevel(logging.DEBUG)
+    log.addHandler(fh)
```
Comment (akshansh237) [inline netmiko/cisco/cisco_vxr_ssh.py]: remove this duplicate addition of handler. also we need to add a check if handler is already added. this is to avoid paramiko logs printed twice in current releases

File: netmiko/cisco/cisco_vxr_ssh.py [hunk 3]
```diff
@@ -0,0 +1,634 @@
...
+class CiscoVxrSSH(CiscoXrSSH):
+    """
+    CiscoVxrSSH is based of CiscoXrSSH -- CiscoBaseConnection
+    """
+
+    def __init__(self, **kwargs):
+        """Constructor
+        """
+        # 30 minutes
+        self.max_read_timeout = kwargs.get('max_read_timeout', 1800)
```
Comment (akshansh237) [inline netmiko/cisco/cisco_vxr_ssh.py]: remove this completely and only use read_timeout. nobody except cafykit connection library passes this, so change there as well

File: netmiko/cisco/cisco_vxr_ssh.py [hunk 4]
```diff
@@ -0,0 +1,634 @@
...
+            # Therefore calling super().write_channel()  which has same functionality as netmiko2's _write_channel
+            # self._write_channel(out_data)
+            super().write_channel(out_data)
+        except socket.error:
+            msg = "Session went down while writing command on channel. Command: {}".format(out_data)
+            log.error(msg)
+            raise SessionDownException(msg)
+        finally:
+            # Always unlock the SSH channel, even on exception.
+            self._unlock_netmiko_session()
```
Comment (akshansh237) [inline netmiko/cisco/cisco_vxr_ssh.py]: is this needed since write_channel() itself has a decorator doing lock, unlock

File: netmiko/cisco/cisco_vxr_ssh.py [hunk 5]
```diff
@@ -0,0 +1,634 @@
...
+
+    '''
```
Comment (akshansh237) [inline netmiko/cisco/cisco_vxr_ssh.py]: is this commented by mistake ?

File: netmiko/base_connection.py [hunk 6]
```diff
@@ -1320,6 +1321,20 @@ def find_prompt(
                     self.write_channel(self.RETURN)
                     time.sleep(sleep_time)
                     prompt = self.read_channel().strip()
+
+                    vxr_pattern = "last login"
```
Comment (akshansh237) [inline netmiko/base_connection.py]: i dont think this is needed as vxr has a overrided find_prompt

File: netmiko/base_connection.py [hunk 7]
```diff
@@ -1320,6 +1321,20 @@ def find_prompt(
                     self.write_channel(self.RETURN)
```
Comment (akshansh237) [inline netmiko/base_connection.py:1371]: i dont think this way of sending "\n" multiple times is correct. a single "\n" should be enough for us to find the prompt. we should use the find_method in vxr_ssh

File: netmiko/cisco/cisco_xr.py [hunk 8]
```diff
@@ -105,36 +114,74 @@ def commit(
         else:
             command_string = "commit"
 
+        if force:
+            command_string = command_string.replace("commit", "commit force")
+        if best_effort:
+            command_string = command_string.replace("commit", "commit best-effort")
+        if replace:
+            command_string = command_string.replace("commit", "commit replace")
+        print(f"command string is {command_string}")
```
Comment (akshansh237) [inline netmiko/cisco/cisco_xr.py]: change to log and add a logger in this file for paramiko logs

File: netmiko/base_connection.py [hunk 9]
```diff
@@ -1634,6 +1649,8 @@ def send_command(
         if normalize:
             command_string = self.normalize_cmd(command_string)
 
+        log.info("In send_command, cmd:{}, read_timeout:{}".format(command_string, read_timeout))
```
Comment (akshansh237) [inline netmiko/base_connection.py:1699]: related to below code, we should change _prompt_handler to raise exception if find_prompt raises exception. this is what we changed in netmiko2 as well

```
        if expect_string is not None:
            search_pattern = expect_string
        else:
            search_pattern = self._prompt_handler(auto_find_prompt)

File: netmiko/cisco/cisco_xr.py [hunk 10]
```diff
@@ -105,36 +114,74 @@ def commit(
...
-            strip_prompt=False,
-            strip_command=False,
-            read_timeout=read_timeout,
-        )
-        if "onfirm" in new_data:
-            output += new_data
+        if replace:
+            new_data = self._send_command_str(
+                command_string,
+                expect_string=r"This commit will replace or remove the entire running configuration",
```
Comment (akshansh237) [inline netmiko/cisco/cisco_xr.py]: i think we should change all the expected pattern everywhere and provide complete patterns instead of either just the start of pattern or end. this is to make sure router actually prints whole pattern and not get stuck. I am trying to be more stricter on XR based on our VXR experience

in this case, it will be below.
RP/0/RP0/CPU0:ios(config)#commit repl
Tue Apr  4 04:03:52.215 UTC

This commit will replace or remove the entire running configuration. This
operation can be service affecting.
Do you wish to proceed? [no]:

File: netmiko/cisco/cisco_xr.py [hunk 11]
```diff
@@ -105,36 +114,74 @@ def commit(
...
+                strip_prompt=False,
+                strip_command=False,
+                read_timeout=read_timeout,
+            )
+            if "This commit will replace or remove the entire running configuration" in new_data:
+                output += new_data
+                new_data = self._send_command_str(
+                    "yes",
+                    expect_string=r"#",
```
Comment (akshansh237) [inline netmiko/cisco/cisco_xr.py]: this should be r'\)#$'

File: netmiko/cisco/cisco_xr.py [hunk 12]
```diff
@@ -105,36 +114,74 @@ def commit(
...
-            output += new_data
+        if replace:
+            new_data = self._send_command_str(
+                command_string,
+                expect_string=r"This commit will replace or remove the entire running configuration",
+                strip_prompt=False,
+                strip_command=False,
+                read_timeout=read_timeout,
+            )
+            if "This commit will replace or remove the entire running configuration" in new_data:
```
Comment (akshansh237) [inline netmiko/cisco/cisco_xr.py]: is this check needed ? i believe is its not found, _send_command_str will raise exception

File: netmiko/cisco/cisco_xr.py [hunk 13]
```diff
@@ -105,36 +114,74 @@ def commit(
...
+                    read_timeout=read_timeout,
+                )
+        else:
+            # IOS-XR might do this:
+            # This could be a few minutes if your config is large. Confirm? [y/n][confirm]
             new_data = self._send_command_str(
-                "y",
-                expect_string=r"#",
+                command_string,
+                expect_string=r"(#|onfirm)",
```
Comment (akshansh237) [inline netmiko/cisco/cisco_xr.py]: should be r'\)#$' and full pattern for second case

File: netmiko/cisco/cisco_xr.py [hunk 14]
```diff
@@ -105,36 +114,74 @@ def commit(
...
                 strip_prompt=False,
                 strip_command=False,
                 read_timeout=read_timeout,
             )
+            if "onfirm" in new_data:
+                output += new_data
+                new_data = self._send_command_str(
+                    "y",
+                    expect_string=r"#",
```
Comment (akshansh237) [inline netmiko/cisco/cisco_xr.py]: change to r'\)#$'

File: netmiko/cisco/cisco_xr.py [hunk 15]
```diff
@@ -105,36 +114,74 @@ def commit(
...
+                    strip_prompt=False,
+                    strip_command=False,
+                    read_timeout=read_timeout,
+                )
         output += new_data
         if error_marker in output:
```
Comment (akshansh237) [inline netmiko/cisco/cisco_xr.py]: this check should be after the alt_error_marker as if user provides "yes" in commit_error_dialog_dict, then commit can still fail

File: netmiko/cisco/cisco_xr.py [hunk 16]
```diff
@@ -105,36 +114,74 @@ def commit(
...
             raise ValueError(f"Commit failed with the following errors:\n\n{output}")
         if alt_error_marker in output:
+            if commit_error_dialog_dict is not None and alt_error_marker in commit_error_dialog_dict:
+                marker_value = commit_error_dialog_dict[alt_error_marker]
+                output += new_data
```
Comment (akshansh237) [inline netmiko/cisco/cisco_xr.py:231]: i see this pattern in this whole method. we get new_data and the next if condition adds it into output. makes it confusing. we should just do output += self._send_command_str(

File: netmiko/cisco/cisco_xr.py [hunk 17]
```diff
@@ -105,36 +114,74 @@ def commit(
...
+                new_data += self._send_command_str(
```
Comment (akshansh237) [inline netmiko/cisco/cisco_xr.py]: this should be new_data = and not += . better to just use output everywhere and remove new_data completely

File: netmiko/cisco/cisco_xr.py [hunk 18]
```diff
@@ -105,36 +114,74 @@ def commit(
...
+                    read_timeout=read_timeout,
+                )
+                output += new_data
+            else:
             # Other commits occurred, don't proceed with commit
-            output += self._send_command_timing_str(
-                "no", strip_prompt=False, strip_command=False
-            )
-            raise ValueError(f"Commit failed with the following errors:\n\n{output}")
+                output += self._send_command_timing_str(
```
Comment (akshansh237) [inline netmiko/cisco/cisco_xr.py:242]: use _send_command_str here as well, timing methods should be avoided as we know what prmpt to expect

File: netmiko/cisco/cisco_xr.py [hunk 19]
```diff
@@ -105,36 +114,74 @@ def commit(
...
+                    "no", strip_prompt=False, strip_command=False
+                )
+                raise ValueError(f"Commit failed with the following errors:\n\n{output}")
```
Comment (akshansh237) [inline netmiko/cisco/cisco_xr.py]: print a log here. log.debug("commit Output: {0}".format(output))
should add in all major methods, send_command, enter_config_mode, check_config_mode, commit, exit_config_mode, etc.

File: netmiko/cisco/cisco_xr.py [hunk 20]
```diff
@@ -166,11 +213,13 @@ def exit_config_mode(self, exit_config: str = "end", pattern: str = "") -> str:
             # Read until we detect either an Uncommitted change or the end prompt
             if not re.search(r"(Uncommitted|#$)", output):
                 output += self.read_until_pattern(pattern=r"(Uncommitted|#$)")
```
Comment (akshansh237) [inline netmiko/cisco/cisco_xr.py]: same comment as above. change to full pattern

File: netmiko/cisco/cisco_xr.py [hunk 21]
```diff
@@ -166,11 +213,13 @@ def exit_config_mode(self, exit_config: str = "end", pattern: str = "") -> str:
...
-            if "Uncommitted changes found" in output:
+            if "Uncommitted" in output:
                 self.write_channel(self.normalize_cmd("no\n"))
                 output += self.read_until_pattern(pattern=r"[>#]")
             if not re.search(pattern, output, flags=re.M):
```
Comment (akshansh237) [inline netmiko/cisco/cisco_xr.py:316]: can you check why is this needed again ?

File: netmiko/cisco/cisco_vxr_ssh.py (hunk 3 above)
Comment (shreyashah) [reply netmiko/cisco/cisco_vxr_ssh.py]: done.

File: netmiko/cisco/cisco_vxr_ssh.py (hunk 4 above)
Comment (shreyashah) [reply netmiko/cisco/cisco_vxr_ssh.py]: good catch, i have removed this now.

File: netmiko/base_connection.py (hunk 6 above)
Comment (shreyashah) [reply netmiko/base_connection.py]: removed

File: netmiko/cisco/cisco_xr.py (hunk 8 above)
Comment (shreyashah) [reply netmiko/cisco/cisco_xr.py]: this was a debugging print, removed it now.

File: netmiko/cisco/cisco_xr.py (hunk 11 above)
Comment (shreyashah) [reply netmiko/cisco/cisco_xr.py]: This is the same code from the origin repo. https://github.com/ktbyers/netmiko/blob/426171102d259d2caebd9e6ac0b78f4ed3cb3758/netmiko/cisco/cisco_xr.py#L124

Comment (shreyashah) [reply netmiko/cisco/cisco_xr.py]: Trying to keep minimal custom changes unless they are absolute necessary.

File: netmiko/cisco/cisco_xr.py (hunk 13 above)
Comment (shreyashah) [reply netmiko/cisco/cisco_xr.py]: Good to have.
as per the original code
https://github.com/ktbyers/netmiko/blob/426171102d259d2caebd9e6ac0b78f4ed3cb3758/netmiko/cisco/cisco_xr.py#L115

File: netmiko/cisco/cisco_xr.py (hunk 14 above)
Comment (shreyashah) [reply netmiko/cisco/cisco_xr.py]: Good to have.
Keeping it consistent with the origin code.

File: netmiko/cisco/cisco_xr.py (hunk 17 above)
Comment (shreyashah) [reply netmiko/cisco/cisco_xr.py]: Good to have.
matches the origin code.

File: netmiko/cisco/cisco_xr.py (hunk 10 above)
Comment (shreyashah) [reply netmiko/cisco/cisco_xr.py]: I understand why you are suggesting this.
This pattern checking is more aligned with the origin netmiko, and would it require to make sure that the complete big string is same for all the pids, XR flavors  and architecture?

File: netmiko/cisco/cisco_xr.py (hunk 14 above)
Comment (shreyashah) [reply netmiko/cisco/cisco_xr.py]: Trying to keep minimal custom changes unless they are absolute necessary.

File: netmiko/cisco/cisco_xr.py (hunk 12 above)
Comment (shreyashah) [reply netmiko/cisco/cisco_xr.py]: which check u mean? at line 136, if "This commit will replace or remove the entire running configuration" is found, then "yes" is sent

File: netmiko/cisco/cisco_xr.py (hunk 15 above)
Comment (shreyashah) [reply netmiko/cisco/cisco_xr.py]: addressed

File: netmiko/cisco/cisco_xr.py (hunk 20 above)
Comment (shreyashah) [reply netmiko/cisco/cisco_xr.py]: Trying to keep minimal custom changes unless they are absolute necessary.

File: netmiko/cisco/cisco_xr.py (hunk 21 above)
Comment (shreyashah) [reply netmiko/cisco/cisco_xr.py:316]: to search for a user provided pattern. This function has no custom additions by us.

File: netmiko/cisco/cisco_xr.py (hunk 18 above)
Comment (shreyashah) [reply netmiko/cisco/cisco_xr.py:242]: this is same as per origin code.

File: netmiko/cisco/cisco_xr.py (hunk 17 above)
Comment (shreyashah) [reply netmiko/cisco/cisco_xr.py]: fixed for new_data = and not +=.
The reason new_data is added to search for pattern for its respective send_command transaction and output is += new_data because in the end of the commit(), outputn is returned and the entire output of all the send transactions are captured with +=

File: netmiko/cisco/cisco_xr.py (hunk 16 above)
Comment (shreyashah) [reply netmiko/cisco/cisco_xr.py:231]: The reason new_data is added to search for pattern for its respective send_command transaction and output is += new_data because in the end of the commit(), outputn is returned and the entire output of all the send transactions are captured with +=

File: netmiko/cisco/cisco_xr.py (hunk 1 above)
Comment (shreyashah) [reply netmiko/cisco/cisco_xr.py:64]: Yes, the multiple inheritance was getting too complex.

File: netmiko/cisco/cisco_vxr_ssh.py (hunk 2 above)
Comment (shreyashah) [reply netmiko/cisco/cisco_vxr_ssh.py]: removed.

File: netmiko/cisco/cisco_vxr_ssh.py (hunk 5 above)
Comment (shreyashah) [reply netmiko/cisco/cisco_vxr_ssh.py]: yes. uncommented it

File: netmiko/base_connection.py (hunk 9 above)
Comment (shreyashah) [reply netmiko/base_connection.py:1699]: in netmiko2 we added a log.info statement when the exception was swallowed 
https://github.com/cafykit/netmiko/blob/7fcc795484e4ccdecc7b3e8b6e365b1cc5c7afb7/netmiko/base_connection.py#L1095

Comment (shreyashah) [reply netmiko/base_connection.py:1699]: added the similar log.info line in netmiko4

Comment (nishantupadhyay123) [reply netmiko/base_connection.py:1699]: I think it is suggested to pick up the changes from cxr_connection.py.
[netmiko/cisco/cisco_vxr_ssh.py](https://github.com/cafykit/netmiko-v4/pull/14/files#diff-c134a175762ced46d28ff2ae0a5f08fc4ef2669c14543bff8e6da37d4cb3a72b) 

the reason is this code is absolutely wrong in base file. I understand that we want to minimize the changes form original netmiko but this is not the way underlaying layers behave..

netmiko -> paramiko -> router.

if the channel is alive and def _send doesnt fail at paramiko that means anything sent will arrive and we must get a pattern.
if router is stuck it will not respond to 1st or nth. 
so , if we send 1 "\n" or a null byte ,  and we wait for response untill session is alive.

File: netmiko/cisco/cisco_xr.py (hunk 10 above)
Comment (nishantupadhyay123) [reply netmiko/cisco/cisco_xr.py]: but are we interested in this str or direct to prompt ?
also, XR cna change the string at will , we need to be really sure we need this kind of check

File: netmiko/base_connection.py (hunk 9 above)
Comment (shreyashah) [reply netmiko/base_connection.py:1699]: created seperate PR for find_prompt() changes to amke it independent https://github.com/cafykit/netmiko-v4/pull/15

File: netmiko/cisco/cisco_xr.py [hunk 22]
```diff
@@ -105,37 +118,113 @@ def commit(
...
+                msg = "Session went down after sending commit replace command"
+                log.error(msg)
+                raise SessionDownException(msg)
+            except PatternNotFoundException:
+                msg = "Prompt not found after sending commit replace command"
+                log.error(msg)
+                raise PatternNotFoundException(msg)
+            
+            try:
+                if "This commit will replace or remove the entire running configuration" in new_data:
```
Comment (akshansh237) [inline netmiko/cisco/cisco_xr.py:171]: this should not be needed. if this was not in new_data, _send_command_str at line 133 would have raised PatternNotFoundException

File: netmiko/cisco/cisco_xr.py (hunk 16 above)
Comment (akshansh237) [reply netmiko/cisco/cisco_xr.py:231]: this is redundant. line 205 already does it

File: netmiko/cisco/cisco_xr.py (hunk 18 above)
Comment (akshansh237) [reply netmiko/cisco/cisco_xr.py:242]: the original code doesn't work in corner cases where router hangs. timing methods should be completely avoided

File: netmiko/cisco/cisco_xr.py (hunk 16 above)
Comment (akshansh237) [reply netmiko/cisco/cisco_xr.py:231]: also, line 207 checks it in output, not new_data. the pattern is not consistent. i see some places are doing output+= send method (conditions below this line of code) and some are using new_data variable. lets keep it consistent everywhere

File: netmiko/cisco/cisco_xr.py [hunk 23]
```diff
@@ -146,33 +235,67 @@ def check_config_mode(
...
+                raise SessionDownException(msg)
+            except PatternNotFoundException:
+                msg = f"Exec Mode Pattern not found after sending config mode exit command: {exit_config}"
+                log.error(msg)
+                raise PatternNotFoundException(msg)
+            if "Uncommitted" in output:
                 self.write_channel(self.normalize_cmd("no\n"))
-                output += self.read_until_pattern(pattern=r"[>#]")
+                try:
+                    output += self.read_until_pattern(pattern=r"[>#]")
```
Comment (akshansh237) [inline netmiko/cisco/cisco_xr.py:307]: the pattern here should be #$

File: netmiko/cisco/cisco_xr.py (hunk 21 above)
Comment (akshansh237) [reply netmiko/cisco/cisco_xr.py:316]: but why is this even needed. after exiting out of config prompt, we expect it to get back exec-prompt. lets delete this code and warn to user passing the parameter that its not used.

File: netmiko/cisco/cisco_xr.py [hunk 24]
```diff
@@ -146,33 +235,67 @@ def check_config_mode(
...
+                    log.error(msg)
+                    raise PatternNotFoundException(msg)
             if not re.search(pattern, output, flags=re.M):
                 output += self.read_until_pattern(pattern=pattern)
+            if skip_check:
+                return output
             if self.check_config_mode():
-                raise ValueError("Failed to exit configuration mode")
+                raise ConfigModeExitError("Failed to exit configuration mode")
+            log.debug("exit_config_mode Output: {0}".format(output))
```
Comment (akshansh237) [inline netmiko/cisco/cisco_xr.py]: use fstring

File: netmiko/base_connection.py [hunk 25]
```diff
@@ -640,15 +641,16 @@ def read_until_pattern(
...
             time.sleep(loop_delay)
 
-        msg = f"""\n\nPattern not detected: {repr(pattern)} in output.
-
-Things you might try to fix this:
-1. Adjust the regex pattern to better identify the terminating string. Note, in
-many situations the pattern is automatically based on the network device's prompt.
-2. Increase the read_timeout to a larger value.
+        if self.remote_conn.closed:
+                msg = f"Session went down while checking for pattern. Search pattern: {repr(pattern)}"
```
Comment (akshansh237) [inline netmiko/base_connection.py:688]: this indentation looks wrong

File: netmiko/base_connection.py [hunk 26]
```diff
@@ -640,15 +641,16 @@ def read_until_pattern(
                 return output
             time.sleep(loop_delay)
 
-        msg = f"""\n\nPattern not detected: {repr(pattern)} in output.
-
-Things you might try to fix this:
-1. Adjust the regex pattern to better identify the terminating string. Note, in
-many situations the pattern is automatically based on the network device's prompt.
-2. Increase the read_timeout to a larger value.
+        if self.remote_conn.closed:
```
Comment (akshansh237) [inline netmiko/base_connection.py]: this check should be added in above while loop as well

File: netmiko/base_connection.py (hunk 7 above)
Comment (akshansh237) [reply netmiko/base_connection.py:1371]: this is still open.

File: netmiko/base_connection.py [hunk 27]
```diff
@@ -1681,17 +1696,19 @@ def send_command(
...
-
-Things you might try to fix this:
-1. Explicitly set your pattern using the expect_string argument.
-2. Increase the read_timeout to a larger value.
-
-You can also look at the Netmiko session_log or debug log for more information.
-
-"""
-            raise ReadTimeout(msg)
+            if self.remote_conn.closed:
```
Comment (akshansh237) [inline netmiko/base_connection.py]: add this check in while loop as well

File: netmiko/base_connection.py [hunk 28]
```diff
@@ -1681,17 +1696,19 @@ def send_command(
...
+                    msg = "Session went down while checking for prompt after sending command. Search pattern: {}".format(
```
Comment (akshansh237) [inline netmiko/base_connection.py:1749]: this indentation is wrong

**Review by akshansh237 — CHANGES_REQUESTED** [review]

File: netmiko/base_connection.py (hunk 26 above)
Comment (nishantupadhyay123) [reply netmiko/base_connection.py]: Shreya, if you look at read_channel (base netmiko) , it is just looping over and not catching any error from paramiko to see if the session is even alive. I think 613 must have a channel / remote_conn alive check for it even make sense to iterate.

File: netmiko/base_connection.py [hunk 29]
```diff
@@ -1516,7 +1528,8 @@ def _prompt_handler(self, auto_find_prompt: bool) -> str:
             try:
                 prompt = self.find_prompt()
             except ValueError:
-                prompt = self.base_prompt
+                    log.info("ValueError encountered from find_prompt() is not re-raised")
```
Comment (nishantupadhyay123) [inline netmiko/base_connection.py:1580]:
```suggestion
                    log.info("ValueError encountered from find_prompt() where prompt is not found , assuming base prompt and continuing ")
```

**Comment by nishantupadhyay123** [comment]
At the end @akshansh237 and @shreyashah  , I think all the debugging from my oct-dec,2022 didnt translate here or between you guys it got lost. I am okay with this PR but the netmiko is slowing us down and whenever and wherever it is necessary we must get to the failure first. 
So, I am okay if this code goes in , but immediate next one should have a while loop or wait must and must be checked to see if session is alive. 
Error pattern should say :
1) rtr session not alive.
2) Rtr session alive but no prompt found. This should execute some traces collection from new channel.
3) any other error.

File: netmiko/cisco/cisco_xr.py (hunk 18 above)
Comment (shreyashah) [reply netmiko/cisco/cisco_xr.py:242]: > At the end @akshansh237 and @shreyashah , I think all the debugging from my oct-dec,2022 didnt translate here or between you guys it got lost. I am okay with this PR but the netmiko is slowing us down and whenever and wherever it is necessary we must get to the failure first. So, I am okay if this code goes in , but immediate next one should have a while loop or wait must and must be checked to see if session is alive. Error pattern should say :
> 
> 1. rtr session not alive.
> 2. Rtr session alive but no prompt found. This should execute some traces collection from new channel.
> 3. any other error.




Both 1 and 2 are added in the code. You can check send_command(),  read_until_pattern() functions in base_connection.py

File: netmiko/cisco/cisco_xr.py (hunk 21 above)
Comment (shreyashah) [reply netmiko/cisco/cisco_xr.py:316]: This if condition exists from the origin code with no custom addition by us. I would keep it there to keep more consistent with the open source code

File: netmiko/base_connection.py [hunk 30]
```diff
@@ -551,6 +552,19 @@ def is_alive(self) -> bool:
...
+        Returns a boolean flag with the state of the connection
+        this is a cafy custom function needed because in cisco_vxr_Ssh code uses self.remote_conn.closed
+        attribute to check connection status. However the 'closed' attriubute is present on remote_conn object only if protocol is ssh
+        and not telnet. Therefore to handle telnet remote_Conn, this function is created a wrapper
+        """
+        if self.protocol == "telnet":
+            return not self.is_alive()
+        else:
+            return self.remote_conn.closed
+
```
Comment (nishantupadhyay123) [inline netmiko/base_connection.py:597]: In next PR ,the code should be 
```suggestion
    def is_session_closed(self) -> bool:
        """
        Returns a boolean flag with the state of the connection
        this is a cafy custom function needed because in cisco_vxr_Ssh code uses self.remote_conn.closed
        attribute to check connection status. However the 'closed' attriubute is present on remote_conn object only if protocol is ssh
        and not telnet. Therefore to handle telnet remote_Conn, this function is created a wrapper
        """
        if self.protocol == "telnet":
            return not self.sock._closed
        else:
            return self.remote_conn.closed
        
```

File: netmiko/base_connection.py (hunk 7 above)
Comment (nishantupadhyay123) [reply netmiko/base_connection.py:1371]: if the VXR code is not impacted then I would suggest to take this seperate. I think it is absolutely not needed.

**Review by nishantupadhyay123 — APPROVED** [review]

**Review by akshansh237 — APPROVED** [review]
